### PR TITLE
Use `promu` for builds and releases

### PR DESCRIPTION
**Pull request by joshuaspence**
Use the `promu` tool for builds and releases. This ensures that this exporter is consistent with other Prometheus exporters, specifically official exporters that live under the `prometheus` namespace.

### CODE REVIEW

**Comment by joshuaspence** [comment]
I'm not entirely sure how `promu release` is meant to work with Travis. I've taken a look at a bunch of other Prometheus exporters and can't see how they are publishing GitHub releases.

**Comment by jonnenauha** [comment]
You just removed my full automation for releases by making a git tag :) Replaced it with a something that does less for me. 

What features promu bring that you are wanting? Only thing I can see is missing certain platform combos. I left a lot of go supported os/arch out because the release page would get cluttered, the main systems should be there.

The packaging is a .tar.gz/.zip file with a single executable inside of it. How complex does this need to be? Is the zip naming convention something that bothers you? Why do I need to be consistent with other prometheus exporter repos? Is it the fact that they can run `make`.

This is not the first time promu is being suggested. Same as last time I would need a reason to do this. Now its even a downgrade cause I lose the release automation (this would have been a fairly trivial fix but you decided to just remove it).

Btw. This is not a official prometheus repo. This is my exporter. It was just submitted to the official docs that list plugins.

**Comment by jonnenauha** [comment]
Also build is failing. Apparently promus make does not run `go get` prior to building or?

File: .promu.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+repository:
+  path: github.com/jonnenauha/prometheus_varnish_exporter
+build:
+  flags: -a -tags netgo
+  ldflags: |
+    -X {{repoPath}}/vendor/github.com/prometheus/common/version.Version={{.Version}}
+    -X {{repoPath}}/vendor/github.com/prometheus/common/version.Revision={{.Revision}}
+    -X {{repoPath}}/vendor/github.com/prometheus/common/version.Branch={{.Branch}}
+    -X {{repoPath}}/vendor/github.com/prometheus/common/version.BuildUser={{user}}@{{host}}
+    -X {{repoPath}}/vendor/github.com/prometheus/common/version.BuildDate={{date "20060102-15:04:05"}}
```
Comment (jonnenauha) [inline .promu.yml]: I do have most of these as well. You can get a version, a git hash and a build date by running with `-version`.

File: .promu.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    - freebsd/386
+    - freebsd/amd64
+    - linux/386
+    - linux/amd64
+    - netbsd/386
+    - netbsd/amd64
+    - openbsd/386
+    - openbsd/amd64
+    - windows/386
+    - windows/amd64
```
Comment (jonnenauha) [inline .promu.yml]: I had all of these as well.

**Comment by jonnenauha** [comment]
The makefile seems kind of nice I suppose. I personally dislike putting monster make files into go projects that are just simple go commands. But if you think that is alienating people I should probably add it.

**Comment by joshuaspence** [comment]
Sorry, I was meaning to come back to this and comment with some additional context. To be honest, it's not the tooling that I care about. I'm not fussed whether this project uses `make`, `promu` or `build.sh`. What I really care about is the naming convention for the release artefacts. Basically, I want to be able to install this plugin using the `prometheus::daemon` resource provided by the [puppet-prometheus](https://github.com/voxpupuli/puppet-prometheus) module. The module seems to work for exporters that were released using `promu`, but it doesn't quite work OOTB for this exporter.

**Comment by joshuaspence** [comment]
Yes, this is currently failing in Travis. It still needs some more work, I just wanted to get some idea of whether or not this PR is likely to be accepted before I spend too much time on it.

**Comment by jonnenauha** [comment]
So it probably assumes certain cmd line params from the exporter that you get by importing promu? Or can it be a stickler for the archive names? That would be pretty crazy.

Well, this is anyway a good reason to investigate :) I will take a look at this daemon and puppet business.

Did you try if it runs now with these changes? I mean there is no new cmd line params etc. Only thing I can think of are the file names (?!).

**Comment by jonnenauha** [comment]
Looks like you just need to create a config file like this https://github.com/voxpupuli/puppet-prometheus/blob/master/manifests/node_exporter.pp and then there are config for the values https://github.com/voxpupuli/puppet-prometheus/blob/master/manifests/params.pp#L45-L53

Looks to be very configurable from the download URL down to the cmd line params. It will create a daemon file for whatever system you need.

I think where it breaks is that their packages have a sub folder inside of them when mine don't. It looks like its assuming that eg. https://github.com/voxpupuli/puppet-prometheus/blob/master/templates/daemon.systemd.erb#L10 

How hard is it to test that whole thing. I have never ran puppet, do I need the full system to use that repo?

**Comment by joshuaspence** [comment]
Yeah the main (only?) Issue was really the structure of the files within the tarball.

**Comment by joshuaspence** [comment]
This is the puppet code I was using for testing, pointing to my fork.

```
  $download_base_url = 'https://github.com/joshuaspence/prometheus_varnish_exporter'
  $download_extension = 'tar.gz'
  $os = $::prometheus::params::os
  $arch = $::prometheus::params::arch

  prometheus::daemon { 'prometheus_varnish_exporter':
    install_method     => 'url',
    version            => $version,
    download_extension => $download_extension,
    os                 => $os,
    arch               => $::prometheus::params::arch,
    real_download_url  => "${download_base_url}/releases/download/v${version}/prometheus_varnish_exporter-${version}.${os}-${arch}.${download_extension}",
    bin_dir            => $::prometheus::params::bin_dir,
    notify_service     => Service['prometheus_varnish_exporter'],
    package_name       => undef,
    package_ensure     => undef,
    manage_user        => true,
    user               => 'varnish-exporter',
    extra_groups       => [],
    group              => 'varnish-exporter',
    manage_group       => true,
    purge              => true,
    options            => '',
    init_style         => $::prometheus::params::init_style,
    service_ensure     => 'running',
    service_enable     => true,
    manage_service     => true,
  }
```

**Comment by jonnenauha** [comment]
Yeah it looks like this is the thing that brakes $binary = "${::staging::path}/${name}-${version}.${os}-${arch}/${name}"

I will fix my release names and its internal structure to mirror whats going on in promu.

**Comment by jonnenauha** [comment]
@joshuaspence can you try https://github.com/jonnenauha/prometheus_varnish_exporter/releases/tag/1.3

Change your config to point to my repo and give it a go.

**Comment by joshuaspence** [comment]
Yep thanks, that works for me.

**Pull request closed without merging**